File: willow_tree_constr/prob.py
import numpy as np
from scipy.stats import norm

def Prob(current_ht, next_ht, alpha, beta, gamma, omega):
    """
    This function calculates the transition probability from current_ht to next_ht.

    Parameters:
    current_ht (numpy.ndarray): Current values of ht
    next_ht (numpy.ndarray): Next values of ht
    alpha (float): Parameter for Heston-Nadi GARCH
    beta (float): Parameter for Heston-Nadi GARCH
    gamma (float): Parameter for Heston-Nadi GARCH
    omega (float): Parameter for Heston-Nadi GARCH

    Returns:
    p (numpy.ndarray): Transition probability from current_ht to next_ht
    """
    next_ht = np.array(next_ht)
    int_points = (next_ht[:-1] + next_ht[1:]) / 2
    print(f"int points {int_points}")
    num_ht = 1
    up_bound = np.nan * np.ones(len(int_points))
    low_bound = np.nan * np.ones(len(int_points))
    
    for i in range(num_ht):
        now_ht = current_ht
       # now_ht = np.full_like(int_points, now_ht)  # Create an array with the same shape as int_points
        up_bound[i] = float(gamma) * np.sqrt(now_ht) + np.sqrt((float(int_points - beta) * now_ht - omega) / alpha)
        low_bound[i] = float(gamma) * np.sqrt(now_ht) - np.sqrt((float(int_points - beta) * now_ht - omega) / alpha)
    
    
    if alpha > 0:
        prob = norm.cdf(up_bound) - norm.cdf(low_bound)
        prob = np.vstack((np.zeros((1, num_ht)), prob, np.ones((1, num_ht))))
        p = np.diff(prob, axis=0).T
    elif alpha < 0:
        prob = 1 - (norm.cdf(up_bound) - norm.cdf(low_bound))
        prob = np.vstack((np.ones((1, num_ht)), prob, np.zeros((1, num_ht))))
        p = (prob[:-1, :] - prob[1:, :]).T

    return p

File: willow_tree_constr/matlab/Prob.m
function p = Prob(current_ht,next_ht, alpha, beta,gamma, omega)
        %================================================
        % 计算概率转移矩阵
        % current_ht是一个列向量（或一个点），next_ht也是一个列向量
        % 返回p是一个行向量（或一个矩阵）
        %================================================
        intPoints = (next_ht(1:end-1)+next_ht(2:end))/2;
        %         intPoints = [-inf;intPoints;inf];
        numHt = length(current_ht);
        upBound = nan(length(intPoints),numHt);
        lowBound = nan(length(intPoints),numHt);
        for i=1:numHt
            nowHt = current_ht(i);
            upBound(:,i) = gamma*sqrt(nowHt)+sqrt((intPoints-beta*nowHt-omega)/alpha);
            lowBound(:,i) = gamma*sqrt(nowHt)-sqrt((intPoints-beta*nowHt-omega)/alpha);
        end
        if alpha>0
            prob = normcdf(real(upBound))-normcdf(real(lowBound));
            prob = [zeros(1,numHt);prob;ones(1,numHt)];
            p = (diff(prob))';
        elseif alpha<0  % alpha可能有限制不可以小于0，需进一步确认
            prob = 1-(normcdf(real(upBound))-normcdf(real(lowBound)));
            prob = [ones(1,numHt);prob;zeroes(1,numHt)];
            p = (prob(1:end-1,:)-prob(2:end,:))';
        end
        
    end